### ISSUE TITLE
Simplify intersection of unions within collections

### ISSUE THREAD

**Issue by Jammf** [astral-sh/ty#3079]
### Summary

Here's a minimal(-ish) example of unexpected type errors involving taking intersections of unions:

```py
from collections.abc import Sequence, Collection, Set
from typing import final
from ty_extensions import Intersection

@final
class X: ...

@final
class Y: ...

def filter1(a: Set[X | Y], b: Set[Y]) -> Set[Intersection[X | Y, Y]]:
    return a & b    # Return type does not match returned value: expected `AbstractSet[Y]`, found `AbstractSet[X | Y]`

def filter2(a: Collection[X | Y], b: Collection[Y]) -> set[Intersection[X | Y, Y]]:
    return {i for i in a if i in b}    # Return type does not match returned value: expected `set[Y]`, found `set[Y | X]`

def filter3(a: Sequence[X | Y], b: Collection[Y]) -> list[Y]:
    return [i for i in a if i in b]    # Return type does not match returned value: expected `list[Y]`, found `list[Y | X]`
```
[Playground](https://play.ty.dev/f471a87d-b246-4c32-8242-b90f77cd8d3e)

I've grouped these together into one issue because they seem to be related, and I also assume the other collection types behave similarly. I wasn't able to find an existing issue that exactly matched this one, though https://github.com/astral-sh/ty/issues/224 seems close.

### Version

ty 0.0.23 (be6e94a27 2026-03-13)

**Comment by carljm** [comment]
Thanks for the report!

The occurrences of `Intersection[X | Y, Y]` in a couple of the return types trivially simplify to just `Y`, and we can simplify by just replacing those with `Y`, without changing the behavior of the example.

The operation `a & b` in `filter1` de-sugars to `a.__and__(b)`, which invokes [this definition](https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L713) of `AbstractSet.__and__` in typeshed:

```py
    def __and__(self, other: AbstractSet[Any], /) -> AbstractSet[_T_co]: ...
```

To support the behavior you are looking for here, this typeshed definition would have to be changed to something like:

```py
    def __and__[S](self, other: MySet[S]) -> MySet[Intersection[S, T]]: ...
```

But this won't happen unless intersection types are standardized in a PEP.

Given that definition, this [already works in ty](https://play.ty.dev/afd42795-470c-4169-9570-ac42e4c086c7).

We could special-case this particular operation in ty, overriding the typeshed definition, but it's not likely to be a high priority to do so.

The other two comprehension cases are complicated by the fact that a subtype of `Collection[Y]` could override `__contains__` to implement arbitrary behavior, so it's technically not sound to assume that `i in b` ensures that `b` is narrowed to the type `Y` -- and we currently don't make that assumption. Neither does any other type checker -- all fail on this example:

```py
from typing import Collection, final

@final
class X: ...

@final
class Y: ...

def _(xy: X | Y, ys: Collection[Y]) -> Y | None:
    if xy in ys:
        return xy  # all type checkers still consider the type of `xy` here to be `X | Y`
    return None
```

I think ty is behaving reasonably and as-designed here, so I'll close this for now as non-actionable. Happy to continue the discussion here, though, if you'd like to propose how ty's behavior should change!
